tsl/src/chunk.c

```diff
@@ -398,103 +454,214 @@ chunk_drop_replica(PG_FUNCTION_ARGS)
-						quote_identifier(chunk->fd.schema_name.data),
-						quote_identifier(chunk->fd.table_name.data));
-	data_nodes = list_make1((char *) node_name);
-	ts_dist_cmd_run_on_data_nodes(drop_cmd, data_nodes, true);
-
-	/*
-	 * This chunk might have this data node as primary, change that association
-	 * if so. Then delete the chunk_id and node_name association.
-	 */
-	chunk_update_foreign_server_if_needed(chunk->fd.id, server->serverid);
-	ts_chunk_data_node_delete_by_chunk_id_and_node_name(chunk->fd.id, node_name);
+	chunk_api_call_chunk_drop_replica(chunk, node_name, server->serverid);
 
 	PG_RETURN_VOID();
 }
 
-typedef struct
-{
-	/* chunk to copy */
-	Chunk *chunk;
-	/* shared name used to name replication slot, publication and
-	 * subscription */
-	NameData operation_id;
-	/* from node */
-	List *src_node;
-	/* to node */
-	List *dst_node;
-} ChunkCopyData;
-
 static void
-chunk_copy_data_prepare(ChunkCopyData *ccd)
+chunk_copy_data_prepare(ChunkMoveData *ccd, bool transactional)
 {
 	const char *cmd;
 	const char *connection_string;
+	NameData application_name;
+
+	if (transactional)
+		StartTransactionCommand();
 
 	/* create publication on the source data node */
 	cmd = psprintf("CREATE PUBLICATION %s FOR TABLE %s",
 				   NameStr(ccd->operation_id),
 				   quote_qualified_identifier(NameStr(ccd->chunk->fd.schema_name),
 											  NameStr(ccd->chunk->fd.table_name)));
-	ts_dist_cmd_run_on_data_nodes(cmd, ccd->src_node, false);
 
-	/* create subscription on the destination data node */
+	/* create the publication in autocommit mode */
+	ts_dist_cmd_run_on_data_nodes(cmd, list_make1((char *) ccd->src_node), false);
+
+	/* track the progress in the catalog */
+	if (transactional)
+	{
+		snprintf(application_name.data,
+				 sizeof(application_name.data),
+				 "publication_created:%s",
+				 ccd->operation_id.data);
+		pgstat_report_appname(application_name.data);
+
+		Assert(ccd->id != 0);
+		ccd->stage = "publication_created";
+		chunk_move_activity_update(ccd);
+
+		CommitTransactionCommand();
+	}
+
+	/* create subscription on the destination data node in a new transaction block */
+	if (transactional)
+		StartTransactionCommand();
 
 	/*
 	 * CREATE SUBSCRIPTION from a database within the same database cluster will hang,
 	 * create the replication slot separately before creating the subscription
 	 */
 	cmd = psprintf("SELECT pg_create_logical_replication_slot('%s', 'pgoutput')",
 				   NameStr(ccd->operation_id));
-	ts_dist_cmd_run_on_data_nodes(cmd, ccd->src_node, false);
+	ts_dist_cmd_run_on_data_nodes(cmd, list_make1((char *) ccd->src_node), false);
+
+	if (transactional)
+	{
+		snprintf(application_name.data,
+				 sizeof(application_name.data),
+				 "replslot_created:%s",
+				 ccd->operation_id.data);
+		pgstat_report_appname(application_name.data);
+
+		Assert(ccd->id != 0);
+		ccd->stage = "replslot_created";
+		chunk_move_activity_update(ccd);
 
+		CommitTransactionCommand();
+	}
```

Comment:
This same code-block seems to duplicated (copied) several times throughout the PR. I suggest moving this into its own function instead of duplicating the same code in many places. At the very least, move the appname update into `chunk_move_activity_update`


Ideally, I think it would be best to think of this as a state machine we move forward as shown in pseudo-code below. (I realize this is a bigger change, but might be worthwhile. We can also refactor later):

```
void
chunk_execute_next_copy_stage(ChunkMoveData *cmd, bool transactional)
{
     const char *next_cmd;
     
     if (transactional)
         StartTransactionCommand();

     switch (cmd->stage)
     {
          case CMA_INIT:
                 cmd->stage = CMA_CREATE_SUB;
                 next_cmd = psprintf(""CREATE SUBSCRIPTION %s CONNECTION '%s' PUBLICATION %s" ....);
                 break;
           case CMA_CREATE_SUB:
            ...
       }

       ts_dist_cmd_run_on_data_nodes(next_cmd, list_make1((char *) ccd->dst_node), false);

       if (transactional)
       {
            // update appname
       }
}
```

Reply:
Should be addressed via the restructuring PR